include/seqan3/range/metafunction.hpp

```diff
@@ -0,0 +1,170 @@
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3
+{
+
+/*!\addtogroup range
+ * \{
+ */
+
+// ----------------------------------------------------------------------------
+
+/*!\brief Whether a type has a `value_type` member and/or `ranges::value_type_t<t>` resolves [Value metafunction].
+ * \tparam t The type to recurse on; must have `ranges::value_type_t<t>`
+ * \hideinitializer
+ */
+template <typename t>
+constexpr bool has_value_type_v = false;
+
+//!\cond
+template <typename t>
+    requires requires (t) { typename ranges::value_type<std::decay_t<t>>::type; }
+constexpr bool has_value_type_v<t> = true;
+//!\endcond
+
+// ----------------------------------------------------------------------------
+
+/*!\brief Recursively determines the `value_type` on containers and/or iterators [Type metafunction].
+ * \tparam t The type to recurse on; must have `ranges::value_type_t<t>`
+ * \hideinitializer
+ */
+template <typename t>
+//!\cond
+    requires has_value_type_v<t>
+//!\endcond
+struct innermost_value_type
+{
+    //!\brief The forwarded type.
+    using type = ranges::value_type_t<std::decay_t<t>>;
+};
+
+//!\cond
+template <typename t>
+    requires has_value_type_v<t> &&
+             has_value_type_v<ranges::value_type_t<std::decay_t<t>>>
+struct innermost_value_type<t>
+{
+    using type = typename innermost_value_type<ranges::value_type_t<std::decay_t<t>>>::type;
+};
+//!\endcond
+
+// ----------------------------------------------------------------------------
+
+//!\brief Shortcut for seqan3::innermost_value_type.
+template <typename t>
+//!\cond
+    requires has_value_type_v<t>
+//!\endcond
+using innermost_value_type_t = typename innermost_value_type<t>::type;
+
+// ----------------------------------------------------------------------------
+
+/*!\brief Returns the number of times you can call `ranges::value_type_t` recursively on t [Value metafunction].
+ * \tparam t The type to be queried; must resolve `ranges::value_type_t` at least once.
+ * \hideinitializer
+ */
+template <typename t>
+//!\cond
+    requires has_value_type_v<t>
+//!\endcond
+constexpr size_t dimension_v = 1;
+
+//!\cond
+template <typename t>
+    requires has_value_type_v<t> &&
+             has_value_type_v<ranges::value_type_t<std::decay_t<t>>>
+constexpr size_t dimension_v<t> = dimension_v<ranges::value_type_t<std::decay_t<t>>> + 1;
+//!\endcond
+
+// ----------------------------------------------------------------------------
+
+/*!\interface seqan3::alike_concept <>
+ * \brief Two types are alike if their seqan3::dimension_v and their decayed seqan3::innermost_value_type_t are
+ * the same.
+ *
+ * \details
+ *
+ * ```cpp
+ * // these evaluate to true:
+ * static_assert(seqan3::alike_concept<std::string,              std::vector<char>>);
+ * static_assert(seqan3::alike_concept<std::vector<std::string>, std::vector<std::vector<char>>>);
+ * ```
+ */
+//!\cond
+template <typename t1, typename t2>
+concept bool alike_concept = requires (t1, t2)
```

Comment:
hey, c'mon, I introduced this, because you complained about the unreadable concept requirements and their lack of documentation (which was a valid criticism).
I am pretty sure this will be useful down-the-road (basically for all of our containers that want some flexibility with assignment from/to other ones).